### PR TITLE
Basic implementation of server architecture in terraform.

### PR DESCRIPTION
**Pull request by yanniboi**

### CODE REVIEW

File: backend/lambda/function.mjs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+                            status: requestJSON.status,
+                            creator: requestData.username,
+                            levelData: requestJSON.levelData,
+                        },
+                    })
+                );
+                body = `Updated level ${event.pathParameters.id}`;
+                break;
+            case "DELETE /levels/{id}":
+                // We dont actually want to delete so lets just unpublish.
```
Comment (HaywardMorihara) [inline backend/lambda/function.mjs:100]: Ah right, yeah I was going to take some time to more thoroughly define what these APIs should be/look like https://github.com/LPGameDevs/EditarrrPublic/issues/53 and probably change them a bit (or maybe a lot)

File: backend/terraform/0-provider.tf [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# Project dependencies
+terraform {
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      version = "~> 4.0"
+    }
+    random = {
+      source  = "hashicorp/random"
```
Comment (HaywardMorihara) [inline backend/terraform/0-provider.tf:9]: What is this being used for?

**Review by HaywardMorihara — APPROVED** [review]
Very cool! I'm know nothing about terraform, but its very human-readable (the comments help) and I see where it matches up with the CloudFormation config

And it seems like I _should_ be able to run this against my own AWS account if I wanted to and it would set up all the resources, yeah?

One downside I see is it doesn't make use of CloudFormation, where one nice thing I've found with that is that keeps all the related resources for the "stack" associated and together in one place so you can easily hop between them and see what is being used. Do you get anything like that out of the box when you use Terraform - maybe via tagging or something?

**Comment by yanniboi** [comment]
> And it seems like I _should_ be able to run this against my own AWS account if I wanted to and it would set up all the resources, yeah?

Yup, you can run this with any aws credentials and it should work. You may want to update the region in 0-providers.

> One downside I see is it doesn't make use of CloudFormation, where one nice thing I've found with that is that keeps all the related resources for the "stack" associated and together in one place so you can easily hop between them and see what is being used. Do you get anything like that out of the box when you use Terraform - maybe via tagging or something?

Tagging is one option, but if you run `terraform destroy` it should get rid of everything as it groups it all together nicely. Maybe you could show me the cloudformation stack at some point? I mean there is also nothing stopping us from addding a cloudformation stack into terraform if you think its helpful :)

File: backend/lambda/function.mjs (hunk 1 above)
Comment (yanniboi) [reply backend/lambda/function.mjs:100]: So I was about to add the dynamodb permission for DELETE and I thought... hmmm maybe we dont want that option? I then went on a brief tangent doing a lot of querying and decided against fixing everything now. We can update/define the requirements later :)

File: backend/terraform/0-provider.tf (hunk 2 above)
Comment (yanniboi) [reply backend/terraform/0-provider.tf:9]: I think its for generating a random string for the s3 bucket. Might not be necessary.

File: backend/terraform/1-lambda-bucket.tf [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  length = 2
+}
+
+# We create the bucket and set force_destroy to true so that terraform can destroy the bucket when we destroy the infrastructure.
+resource "aws_s3_bucket" "lambda_bucket" {
+  bucket        = random_pet.lambda_bucket_name.id
+  force_destroy = true
+}
+
+# We set the public access block to true so that the bucket is not public.
```
[the remark this answers is not in the document]
Comment (yanniboi) [reply backend/terraform/1-lambda-bucket.tf:15]: LOL! Github copilot doing my documentation for me and me not correcting it properly :rofl:

File: backend/terraform/0-provider.tf (hunk 2 above)
Comment (HaywardMorihara) [reply backend/terraform/0-provider.tf:9]: Ahh I saw that, interesting. Thanks!

Comment (HaywardMorihara) [reply backend/terraform/0-provider.tf:9]:
```suggestion
      # Generates a random string for the S3 bucket
      source  = "hashicorp/random"
```

File: backend/terraform/1-lambda-bucket.tf (hunk 3 above)
[the remark this answers is not in the document]
Comment (HaywardMorihara) [reply backend/terraform/1-lambda-bucket.tf:15]:
```suggestion
# We set the public access block to true so that the bucket is public.
```

**Comment by HaywardMorihara** [comment]
> Yup, you can run this with any aws credentials and it should work. You may want to update the region in 0-providers.

Sweet! I will try it out.

> Tagging is one option, but if you run `terraform destroy` it should get rid of everything as it groups it all together nicely. Maybe you could show me the cloudformation stack at some point? I mean there is also nothing stopping us from addding a cloudformation stack into terraform if you think its helpful :)

Yeah, but it isn't very important. It's just kinda nice how you can go into the AWS console and see everything together (example in screenshot):
Screen Shot 2023-06-07 at 6 34 02 PM

**Comment by yanniboi** [comment]
> > Yup, you can run this with any aws credentials and it should work. You may want to update the region in 0-providers.
> 
> Sweet! I will try it out.
> 
> > Tagging is one option, but if you run `terraform destroy` it should get rid of everything as it groups it all together nicely. Maybe you could show me the cloudformation stack at some point? I mean there is also nothing stopping us from addding a cloudformation stack into terraform if you think its helpful :)
> 
> Yeah, but it isn't very important. It's just kinda nice how you can go into the AWS console and see everything together (example in screenshot): Screen Shot 2023-06-07 at 6 34 02 PM

That is nice 👌

**Pull request merged**